shared/src/main/scala/mlscript/NuTypeDefs.scala

```diff
@@ -460,6 +470,71 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
   }
   
   
+  def getRefs(body: Statement): RefMap = {
+    val refs = mutable.HashSet[(Str, Bool)]()
+
+    def visit(s: Statement): Bool = s match {
+      case App(lhs, rhs) => visit(lhs) || visit(rhs)
+      case CaseOf(trm, cases) =>
+        def visitCases(cases: CaseBranches): Bool = cases match {
+          case Case(pat, body, rest) => visit(pat) || visit(body) || visitCases(cases)
+          case Wildcard(body) => visit(body)
+          case NoCases => false
+        }
+        visit(trm) || visitCases(cases)
+      case Asc(trm, ty) => visit(trm)
+      case Assign(lhs, rhs) => visit(lhs) || visit(rhs)
+      case Bind(lhs, rhs) => visit(lhs) || visit(rhs)
+      case Blk(stmts) => stmts.foldLeft(false)((r, s) => r || visit(s))
+      case Bra(_, trm) => visit(trm)
+      case Constructor(params, body) => visit(body)
+      case Eqn(lhs, rhs) => visit(lhs) || visit(rhs)
+      case Forall(params, body) => visit(body)
+      case If(body, els) =>
+        def visitIfBody(body: IfBody): Bool = body match {
+          case IfThen(expr, rhs) => visit(expr) || visit(rhs)
+          case IfElse(body) => visit(body)
+          case IfLet(_, _, rhs, body) => visit(rhs)
+          case IfOpApp(lhs, _, rhs) => visit(lhs)
+          case IfOpsApp(lhs, opsRhss) => visit(lhs)
+          case IfBlock(lines) => lines.foldLeft(false)((r, e) => r || (e match {
+            case L(bd) => visitIfBody(bd)
+            case R(s) => visit(s)
+          }))
+        }
+        visitIfBody(body) || els.fold(false)(t => visit(t))
+      case Inst(body) => visit(body)
+      case Lam(_, rhs) => visit(rhs)
+      case Let(_, name, rhs, body) => visit(rhs) || visit(body)
+      case LetS(_, pat, rhs) => visit(pat) || visit(rhs)
+      case New(head, body) => body.entities.foldLeft(false)((r, s) => r || visit(s))
+      case NuFunDef(_, nme, _, L(rhs)) => visit(rhs)
+      case Rcd(fields) => fields.foldLeft(false)((r, s) => r || visit(s._2.value))
+      case Sel(Var("this"), Var(name)) =>
+        refs.add((name, true))
+        false
+      case Sel(receiver, fieldName) => visit(receiver) || visit(fieldName)
+      case Splc(fields) => fields.foldLeft(false)((r, e) => r || (e match {
+        case L(e) => visit(e)
+        case R(f) => visit(f.value)
+      }))
+      case Subs(arr, idx) => visit(arr) || visit(idx)
+      case Test(trm, ty) => visit(trm) || visit(ty)
+      case Tup(fields) => fields.foldLeft(false)((r, s) => r || visit(s._2.value))
+      case TyApp(lhs, targs) => visit(lhs)
+      case Var(name) =>
+        if (name === "this") true
+        else {
+          refs.add((name, false))
+          false
+        }
+      case Where(body, where) => where.foldLeft(visit(body))((r, s) => r || visit(s))
+      case With(trm, fields) => visit(trm) || visit(fields)
```

Comment:
It would be better to add a `childrenTerms` to `TermImpl`, which could be used here and possibly reused elsewhere, and would avoid this huge pattern matching.

Reply:
Actually, we have a similar code for this:
```scala
trait Located {
  def children: List[Located]
  // ...
}
```

So far I replaced the huge pattern matching with a `foldLeft` on the `children`.

Reply:
Yeah I know, but this traverses every single located thing as opposed to every term. But it's okay to use this for now. (You can close this.)